validator/client/src/main/java/tech/pegasys/teku/validator/client/DoppelgangerDetection.java

```diff
@@ -0,0 +1,288 @@
+                    "No validators doppelganger detected after 2 epochs. Stopping doppelganger detection.");
+                return stopDoppelgangerDetection(new HashMap<>());
+              }
+
+              LOGGER.info(
+                  "Performing a validators doppelganger check for epoch {}, slot {}",
+                  currentEpoch,
+                  currentSlot);
+
+              return validatorIndexProvider
+                  .getValidatorIndices()
+                  .thenCompose(
+                      validatorIndices ->
+                          checkValidatorsLivenessAtEpoch(
+                              currentSlot, currentEpoch, validatorIndices))
+                  .orTimeout(checkDelay)
+                  .exceptionally(
+                      throwable -> {
+                        LOGGER.error(
+                            "Unable to check validators doppelganger. Unable to get validators indices: {}",
+                            extractErrorMessage(throwable));
+                        return null;
+                      })
+                  .thenApply(o -> null);
+            })
+        .orTimeout(checkDelay)
+        .exceptionally(
+            throwable -> {
+              LOGGER.error(
+                  "Unable to check validators doppelganger. Unable to get genesis time to calculate the current epoch: {}",
+                  extractErrorMessage(throwable));
+              return null;
+            });
+  }
+
+  private SafeFuture<Void> checkValidatorsLivenessAtEpoch(
+      UInt64 currentSlot, UInt64 currentEpoch, IntCollection validatorIndices) {
+    return validatorApiChannel
+        .getValidatorsLiveness(mapToUIntIndices(validatorIndices), currentEpoch)
+        .thenAccept(
+            validatorLivenessAtEpoches ->
+                checkValidatorDoppelgangers(
+                    validatorLivenessAtEpoches,
+                    mapToUIntIndices(validatorIndices),
+                    currentEpoch,
+                    currentSlot))
+        .orTimeout(checkDelay)
+        .exceptionally(
+            throwable -> {
+              LOGGER.error(
+                  "Unable to check validators doppelganger. Unable to get validators liveness: {}",
+                  extractErrorMessage(throwable));
+              return null;
+            })
+        .thenApply(doppelgangerDetected -> null);
+  }
+
+  private void checkValidatorDoppelgangers(
+      final Optional<List<ValidatorLivenessAtEpoch>> validatorLivenessAtEpoches,
+      final List<UInt64> validatorIndices,
+      final UInt64 epoch,
+      final UInt64 slot) {
+    List<ValidatorLivenessAtEpoch> liveValidators =
+        filterLiveValidators(validatorLivenessAtEpoches, validatorIndices);
+    if (!liveValidators.isEmpty()) {
+      LOGGER.fatal("Doppelganger detected. Shutting down Validator Client.");
+      this.validatorIndexProvider
+          .getValidatorIndicesByPublicKey()
+          .thenApply(
+              blsPublicKeyIntegerMap -> {
```

Comment:
the default coming from intelliJ sometime can be improved :)

Reply:
Done